File: pycript/temp_file.py
import os
import tempfile
import random
import string

user_home = os.path.expanduser("~")
pycript_dir = os.path.join(user_home, ".pycript")

def parse_temp_file_output(original_data, original_header, temp_file_path):
    with open(temp_file_path, 'rb') as temp_file:
        file_content = temp_file.read()
    body_end_marker = b'\n--BODY_END--\n'

    # Split the file content using the marker
    parts = file_content.split(body_end_marker, 1)

    # Extract body data
    body_data = parts[0]

    # Extract header data if present, otherwise use the original header
    if len(parts) > 1 and parts[1].strip():  # Check if header exists and is not empty
        header_data = parts[1].strip()
    else:
        header_data = original_header
    return body_data, header_data


def create_temp_file(data, headervalue=None):
    # Get the temp directory path
    random_file_name = ''.join([random.choice(string.ascii_letters + string.digits) for _ in range(12)])
    file_path = os.path.join(pycript_dir, random_file_name)
    
    # Write data to the file
    with open(file_path, "wb") as file:
        file.write(bytes(data))  # Write the byte array directly to the file
        file.write(b'\n--BODY_END--\n')  # Write the binary body end marker
        if headervalue is not None:
            file.write(headervalue.encode('utf-8'))
    
    return file_path


def create_temp_dir():
    
    if not os.path.exists(pycript_dir):
        os.makedirs(pycript_dir)
    temp_dir = tempfile.mkdtemp(dir=pycript_dir)
    

File: src/main/java/com/pycript/EncDec/TempFile.java
package com.pycript.EncDec;

import java.io.*;
import java.nio.file.*;
import java.util.Random;
import org.apache.commons.lang3.tuple.Pair;
import org.apache.commons.lang3.tuple.ImmutablePair;
import com.pycript.ui.LogTab;

public class TempFile {

    private static final String PYCRIPT_DIR;
    private static final byte[] BODY_END_MARKER = "\n--BODY_END--\n".getBytes();

    static {
        String userHome = System.getProperty("user.home");
        PYCRIPT_DIR = userHome + File.separator + ".pycript";
        createTempDir();
    }

    public static Pair<byte[], String> parseTempFileOutput(byte[] originalData, String originalHeader, String tempFilePath) {
        try {
            byte[] fileContent = Files.readAllBytes(Paths.get(tempFilePath));

            LogTab logTab = LogTab.getInstance();
            if (logTab != null && logTab.isLoggingEnabled()) {
                logTab.appendLog(new String(fileContent));
            }

            int markerIndex = indexOf(fileContent, BODY_END_MARKER);

            byte[] bodyData;
            String headerData;

            if (markerIndex != -1) {
                bodyData = new byte[markerIndex];
                System.arraycopy(fileContent, 0, bodyData, 0, markerIndex);

                int headerStart = markerIndex + BODY_END_MARKER.length;
                if (headerStart < fileContent.length) {
                    byte[] headerBytes = new byte[fileContent.length - headerStart];
                    System.arraycopy(fileContent, headerStart, headerBytes, 0, headerBytes.length);
                    String extractedHeader = new String(headerBytes).trim();
                    headerData = extractedHeader.isEmpty() ? originalHeader : extractedHeader;
                } else {
                    headerData = originalHeader;
                }
            } else {
                bodyData = fileContent;
                headerData = originalHeader;
            }

            return new ImmutablePair<>(bodyData, headerData);

        } catch (IOException e) {
            return new ImmutablePair<>(originalData, originalHeader);
        }
    }

    public static String createTempFile(byte[] data, String headerValue) throws IOException {
        // #20
        String randomFileName = generateRandomFileName(12);
        String filePath = PYCRIPT_DIR + File.separator + randomFileName;

        try (FileOutputStream fos = new FileOutputStream(filePath)) {
            fos.write(data);

            fos.write(BODY_END_MARKER);

            if (headerValue != null && !headerValue.isEmpty()) {
                fos.write(headerValue.getBytes("UTF-8"));
            }
        }

        return filePath;
    }

    public static void createTempDir() {
        File dir = new File(PYCRIPT_DIR);
        if (!dir.exists()) {
            dir.mkdirs();
        }
    }

    public static void deleteTempFolder() {
        File dir = new File(PYCRIPT_DIR);
        if (dir.exists()) {
            deleteDirectory(dir);
        }
    }

    private static void deleteDirectory(File directory) {
        File[] files = directory.listFiles();
        if (files != null) {
            for (File file : files) {
                if (file.isDirectory()) {
                    deleteDirectory(file);
                } else {
                    file.delete();
                }
            }
        }
        directory.delete();
    }

    private static String generateRandomFileName(int length) {
        String chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
        Random random = new Random();
        StringBuilder sb = new StringBuilder(length);
        for (int i = 0; i < length; i++) {
            sb.append(chars.charAt(random.nextInt(chars.length())));
        }
        return sb.toString();
    }

    private static int indexOf(byte[] array, byte[] pattern) {
        for (int i = 0; i <= array.length - pattern.length; i++) {
            boolean found = true;
            for (int j = 0; j < pattern.length; j++) {
                if (array[i + j] != pattern[j]) {
                    found = false;
                    break;
                }
            }
            if (found) {
                return i;
            }
        }
        return -1;
    }
}
